File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/site/sites/Lainchan.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.core.site.sites;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.github.k1rakishou.chan.core.site.ChunkDownloaderSiteProperties;
import com.github.k1rakishou.chan.core.site.Site;
import com.github.k1rakishou.chan.core.site.SiteIcon;
import com.github.k1rakishou.chan.core.site.common.CommonSite;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanActions;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanApi;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanCommentParser;
import com.github.k1rakishou.chan.core.site.common.vichan.VichanEndpoints;
import com.github.k1rakishou.chan.core.site.parser.CommentParserType;
import com.github.k1rakishou.common.DoNotStrip;
import com.github.k1rakishou.model.data.board.ChanBoard;
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor;
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor;
import com.github.k1rakishou.model.data.descriptor.SiteDescriptor;

import org.jetbrains.annotations.NotNull;

import okhttp3.HttpUrl;

@DoNotStrip
public class Lainchan extends CommonSite {
    private final ChunkDownloaderSiteProperties chunkDownloaderSiteProperties;

    public static final String SITE_NAME = "Lainchan";
    public static final SiteDescriptor SITE_DESCRIPTOR = SiteDescriptor.Companion.create(SITE_NAME);

    public static final CommonSiteUrlHandler URL_HANDLER = new CommonSiteUrlHandler() {
        private static final String ROOT = "https://lainchan.org/";

        @Override
        public Class<? extends Site> getSiteClass() {
            return Lainchan.class;
        }

        @Override
        public HttpUrl getUrl() {
            return HttpUrl.parse(ROOT);
        }

        @Override
        public HttpUrl[] getMediaHosts() {
            return new HttpUrl[]{getUrl()};
        }

        @Override
        public String[] getNames() {
            return new String[]{"lainchan"};
        }

        @Override
        public String desktopUrl(ChanDescriptor chanDescriptor, @Nullable Long postNo) {
            if (chanDescriptor instanceof ChanDescriptor.CatalogDescriptor) {
                return getUrl().newBuilder().addPathSegment(chanDescriptor.boardCode()).toString();
            } else if (chanDescriptor instanceof ChanDescriptor.ThreadDescriptor) {
                return getUrl().newBuilder()
                        .addPathSegment(chanDescriptor.boardCode())
                        .addPathSegment("res")
                        .addPathSegment(((ChanDescriptor.ThreadDescriptor) chanDescriptor).getThreadNo() + ".html")
                        .toString();
            } else {
                return null;
            }
        }
    };

    public Lainchan() {
        chunkDownloaderSiteProperties = new ChunkDownloaderSiteProperties(true, true);
    }

    @Override
    public void setup() {
        setEnabled(true);
        setName(SITE_NAME);
        setIcon(SiteIcon.fromFavicon(getImageLoaderV2(), HttpUrl.parse("https://lainchan.org/favicon.ico")));

        setBoards(
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "λ"), "Programming"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "Δ"), "Do It Yourself"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "sec"), "Security"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "Ω"), "Technology"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "inter"), "Games and Interactive Media"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "lit"), "Literature"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "music"), "Musical and Audible Media"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "vis"), "Visual Media"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "hum"), "Humanity"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "drug"), "Drugs 3.0"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "zzz"), "Consciousness and Dreams"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "layer"), "layer"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "q"), "Questions and Complaints"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "r"), "Random"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "lain"), "Lain"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "culture"), "Culture 15 freshly bumped threads"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "psy"), "Psychopharmacology 15 freshly bumped threads"),
                ChanBoard.create(BoardDescriptor.create(siteDescriptor().getSiteName(), "mega"), "15 freshly bumped threads")
        );

        setResolvable(URL_HANDLER);

        setConfig(new CommonConfig() {
            @Override
            public boolean siteFeature(SiteFeature siteFeature) {
                return super.siteFeature(siteFeature) || siteFeature == SiteFeature.POSTING;
            }
        });

        setEndpoints(new VichanEndpoints(this, "https://lainchan.org", "https://lainchan.org"));
        setActions(new VichanActions(this, getProxiedOkHttpClient(), getSiteManager(), getReplyManager()));
        setApi(new VichanApi(getSiteManager(), getBoardManager(), this));
        setParser(new VichanCommentParser());
    }

    @NotNull
    @Override
    public CommentParserType commentParserType() {
        return CommentParserType.VichanParser;
    }

    @NonNull
    @Override
    public ChunkDownloaderSiteProperties getChunkDownloaderSiteProperties() {
        return chunkDownloaderSiteProperties;
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/core/site/sites/vichan/lainchan/Lainchan.kt
package com.github.k1rakishou.chan.core.site.sites.vichan.lainchan

import com.github.k1rakishou.chan.core.site.SiteActions
import com.github.k1rakishou.chan.core.site.SiteConfiguration
import com.github.k1rakishou.chan.core.site.SiteUrlHandler
import com.github.k1rakishou.chan.core.site.common.DefaultPostParser
import com.github.k1rakishou.chan.core.site.common.vichan.LainchanCommentParser
import com.github.k1rakishou.chan.core.site.limitations.ConstantAttachablesCount
import com.github.k1rakishou.chan.core.site.limitations.ConstantMaxTotalSizeInfo
import com.github.k1rakishou.chan.core.site.limitations.PostingLimitationConfig
import com.github.k1rakishou.chan.core.site.parser.PostParser
import com.github.k1rakishou.chan.core.site.sites.vichan.BaseVichanSite
import com.github.k1rakishou.model.data.board.ChanBoard
import com.github.k1rakishou.model.data.descriptor.BoardDescriptor
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor

class Lainchan : BaseVichanSite(
  defaultDomain = "https://lainchan.org"
) {
  private val boards by lazy {
    buildList {
      val siteName = descriptor.siteName

      add(ChanBoard.create(BoardDescriptor.create(siteName, "λ"), "Programming"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "Δ"), "Do It Yourself"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "sec"), "Security"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "Ω"), "Technology"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "inter"), "Games and Interactive Media"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "lit"), "Literature"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "music"), "Musical and Audible Media"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "vis"), "Visual Media"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "hum"), "Humanity"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "drug"), "Drugs 3.0"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "zzz"), "Consciousness and Dreams"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "layer"), "layer"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "q"), "Questions and Complaints"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "r"), "Random"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "lain"), "Lain"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "culture"), "Culture 15 freshly bumped threads"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "psy"), "Psychopharmacology 15 freshly bumped threads"))
      add(ChanBoard.create(BoardDescriptor.create(siteName, "mega"), "15 freshly bumped threads"))
    }
  }

  override val enabled: Boolean = true
  override val name: String = SITE_NAME
  override val postingLimitationConfig by lazy {
    PostingLimitationConfig(
      postMaxAttachables = ConstantAttachablesCount(3),
      postMaxAttachablesTotalSize = ConstantMaxTotalSizeInfo(75 * (1024 * 1024)) // 75 MB
    )
  }
  override val urlHandler: SiteUrlHandler by lazy { LainchanUrlHandler(this) }
  override val actions: SiteActions by lazy { LainchanActions(this) }
  override val postParser: PostParser by lazy { DefaultPostParser(LainchanCommentParser(), archivesManager) }
  override val staticBoards: List<ChanBoard> = boards

  override fun hasSiteFeature(siteFeature: SiteConfiguration.SiteFeature): Boolean {
    return super.hasSiteFeature(siteFeature) ||
      siteFeature === SiteConfiguration.SiteFeature.Posting
  }

  class LainchanUrlHandler(lainchan: Lainchan) : CommonSiteUrlHandler(lainchan) {
    override fun desktopUrl(chanDescriptor: ChanDescriptor, postNo: Long?, postSubNo: Long?): String? {
      return when (chanDescriptor) {
        is ChanDescriptor.CatalogDescriptor -> {
          rootUrl.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .toString()
        }

        is ChanDescriptor.ThreadDescriptor -> {
          rootUrl.newBuilder()
            .addPathSegment(chanDescriptor.boardCode())
            .addPathSegment("res")
            .addPathSegment(chanDescriptor.threadNo.toString() + ".html")
            .toString()
        }

        else -> null
      }
    }
  }

  companion object {
    const val SITE_NAME = "Lainchan"
  }
}